File: week-2/most-common-word/most-common-word.js
/**
 * @param {string} paragraph
 * @param {string[]} banned
 * @return {string}
 */
var mostCommonWord = function(paragraph, banned) {
    paragraph = paragraph.replace(/,|!|\?|'|,|;|\./g, ' ').toLowerCase().split(/\s+/g);
    const banned_list = {};
    for (let word of banned) {
        banned_list[word] = true;
    }
    const word_counts = {};
    let max_count = 0;
    let max_word = '';
    for (let word of paragraph) {
        if (!(word in banned_list)) {
            word_counts[word] = word in word_counts ? word_counts[word] + 1 : 1;
            if (word_counts[word] > max_count) {
                max_word = word;
                max_count = word_counts[word];
            }
        }
    }
    
    return max_word;
};

File: week-2/most-common-word/most-common-word.py
class Solution:
    def mostCommonWord(self, paragraph: str, banned: List[str]) -> str:
        for i in "!?',;.":
            paragraph = paragraph.replace(i, ' ')
        paragraph = paragraph.lower().split()
        banned_table = set(banned)
        word_counts = {}
        
        max_count = 0
        max_word = None
        for word in paragraph:
            if word not in banned_table:
                word_counts[word] = word_counts[word] + 1 if word in word_counts else 1
                if word_counts[word] > max_count:
                    max_count = word_counts[word]
                    max_word = word
            
        return max_word
